### ISSUE TITLE
Sandboxed Play Services only partly working with Chromecast Casting

### ISSUE THREAD

**Issue by eggplanting** [GrapheneOS/os-issue-tracker#631]
**My situation and what I've done:**
- Running the stable build RQ3A.210805.001.A1.2021081822 on my Pixel 4a
- Successfully installed Sandboxed Play Services like mentioned in this guide: https://grapheneos.org/usage#sandboxed-play-services
- Tried multiple apps which provide a native Chromecast Casting functionality (Youtube, Tidal)
- All of my chromecasts are discovered when Play Services are installed, clicking on them doesn't provide Casting
- Youtube does nothing when a Chromecast is selected, Tidal crashes each time

**Issue:**
Chromecast functionality seems to be party working, since all of my Chromecasts are discovered (not the case without them or with MicroG for example). It seems that some libraries are missing, because nothing happens / some apps are crashing when a Chromecast ist selected.

Would be nice to see if a fully working Chromecast implementation would be possible in future.

Please tell me how I can provide more. Best regards!

**Comment by thestinger** [comment]
It's unlikely that this is going to be able to work since we aren't going to be providing special integration into the OS where the OS uses it as a trusted backend. Unless casting is a generic feature available to any apps, there isn't a way for us to provide it for Play services.

VLC has their own implementation of Chromecast. There may be other apps offering the same thing. It can't provide features requiring privileged OS integration though.

**Comment by thestinger** [comment]
So, unfortunately, I don't think we can provide this unless the APIs it uses are made available to generic apps without special privileges. In that case, even if it didn't properly ask for permission, etc. we could theoretically get it working with shims. As is, I don't really think this is realistic.

**Comment by Simbaclaws** [comment]
Are there any other casting protocols that can be implemented as a alternative protocol for casting apps?

Or can I perhaps use something like a raspberry pi with wifi instead?

**Comment by flawedworld** [comment]
We will likely end up implementing shims for Google Cast in the compatibility layer,

**Comment by thestinger** [comment]
Please test again how much is working on the latest release.

**Comment by Simbaclaws** [comment]
The latest release seems to work with casting. Makes me really happy. Thank you <3

**Comment by thestinger** [comment]
If there's still an issue, let us know.

**Comment by SapientHetero** [comment]
I still have an issue. Latest release running on Pixel 5a sometimes locates devices to cast to, but more often does not, with no discernable pattern (so far, anyway). When devices are located, casting works fine.

**Comment by Zanthed** [comment]
@SapientHetero It isn't related to this issue. As long as casting itself works then this can stay closed.

You're describing a different unrelated issue which can be caused by lots of different things.

**Comment by SapientHetero** [comment]
Should I open a new issue for the connection problem? I'm on the same network as the devices I want to connect to, and not using VPN. Any suggestions for further data collection that might help narrow this down?

------- Original Message -------

**Comment by haxwagon** [comment]
I can now cast from apps which is awesome!  However, screen cast never detects any devices (started from quick settings tile). Should that be working as part of this update, too?

**Comment by thestinger** [comment]
No, that's inherently privileged functionality. It won't even be possible for it to work until that becomes available as an unprivileged feature apps can request, if ever.
